src/BodyComponent.h

```diff
@@ -0,0 +1,212 @@
+	};
+
+	// Type-specific serialization implementation. Delegates to the type's
+	// internal serialization methods and provides the needed glue code.
+	//
+	// The Component::LoadFromJson method will be called after the component
+	// is constructed and added to the body, and may potentially have defaults
+	// set by the owning Body before it is deserialized.
+	template <typename T>
+	struct Serializer final : public SerializerBase {
+		Serializer(std::string name, Pool<T> *pool) :
+			SerializerBase(name),
+			pool(pool)
+		{}
+		Pool<T> *pool;
+
+		virtual void toJson(const Body *body, Json &obj, Space *space) override
+		{
+			pool->get(body)->SaveToJson(obj, space);
+		}
+
+		virtual void fromJson(Body *body, const Json &obj, Space *space) override
+		{
+			auto *component = pool->newComponent(body);
+			component->LoadFromJson(obj, space);
+		}
+	};
+
+	// Returns (and creates) a type-specific pool.
+	template <typename T>
+	static Pool<T> *GetComponentType()
+	{
+		auto iter = m_componentPools.find(TypeId<T>::Get());
+		if (iter == m_componentPools.end()) {
+			auto *pool = new Pool<T>(m_componentIdx++, TypeId<T>::Get());
+			iter = m_componentPools.emplace(TypeId<T>::Get(), pool).first;
+			m_componentTypes.push_back(pool);
+		}
+
+		return static_cast<Pool<T> *>(iter->second.get());
+	}
+
+	// Returns (if present) the polymorphic interface to component associated with the given index
+	// This differs from the type-ID and is volatile between program restarts
+	static PoolBase *GetComponentType(size_t componentIndex)
+	{
+		assert(componentIndex < m_componentTypes.size());
+		return m_componentTypes[componentIndex];
+	}
+
+	// Returns (if present) the polymorphic interface to the component type associated with the given name
+	// This name can be used for serialization or to interface with Lua
+	static PoolBase *GetComponentType(const std::string &typeName)
+	{
+		auto iter = m_componentNames.find(typeName);
+		if (iter != m_componentNames.end())
+			return iter->second;
+
+		return nullptr;
+	}
+
+	// Register a component type to be queryable from Lua with body:GetComponent()
+	template <typename T>
+	static bool RegisterLuaInterface(std::string typeName)
+	{
+		Pool<T> *pool = GetComponentType<T>();
+		if (pool->luaInterface)
+			return false;
+
+		pool->luaInterface = new LuaInterface<T>(pool);
+		m_componentNames.emplace(typeName, pool);
+		return true;
+	}
+
+	// Register a serializer for the given type.
+	// Multiple serializers can be registered for the given type and used while
+	// loading for backwards compatibility, however only the last-registered
+	// serializer will be used when serializing to JSON
+	template <typename T>
+	static bool RegisterSerializer(std::string typeName)
+	{
+		assert(!m_componentSerializers.count(typeName));
+		Pool<T> *pool = GetComponentType<T>();
+
+		SerializerBase *serial = new Serializer<T>(typeName, pool);
+		pool->serializer = serial;
+
+		m_componentSerializers.emplace(typeName, serial);
```

Comment:
Oh no, this is put in `unique_ptr`, but it is also put as a raw pointer in the `pool->serializer`?

Reply:
The unique_ptr is just so the memory is automatically cleaned up at program termination - as far as the system is concerned this can never be a dangling pointer.

Reply:
And specifically, the use of unique_ptr is because there is a many-to-one relationship between serializers and pools; you can register multiple serializers to handle the case in which a component is migrated to a different name or different layout (although there's not currently a good way to implement that at current) so there may be multiple serializers but only one referenced from the pool itself.

Reply:
Added a comment to denote the pointed-to object's lifetime.